SYCL/ESIMD/api/functional/common.hpp

```diff
@@ -0,0 +1,62 @@
+//===-- common.hpp - This file provides common functions for simd constructors
+//      tests -------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Common file for tests on simd class.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+#include "../../esimd_test_utils.hpp"
+#include "logger.hpp"
+#include "type_coverage.hpp"
+#include "type_traits.hpp"
+#include "value.hpp"
+
+#include <vector>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+namespace details {
+
+// Bitwise comparison for two values
+template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
+  constexpr size_t size{sizeof(T)};
+
+  // Such type-punning is OK from the point of strict aliasing rules
+  const auto &lhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(lhs);
+  const auto &rhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(rhs);
+
+  bool result{true};
+  for (size_t i = 0; i < size; ++i) {
+    result &= lhs_bytes[i] == rhs_bytes[i];
+  }
+  return result;
+}
+
+} // namespace details
+
+// A wrapper to speed-up bitwise comparison
+template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
```

Comment:
Right.. Looks like a misprint, the original intent was to limit the bitwise comparison for:

- signed integer type - due to 3 representations available
- floating point type - due to nan values with opcode

Reply:
I understand why detail::are_bitwise_equal() may be needed for float (because of nan, etc), but why can't you use regular operation == for signed integer types?

Reply:
What if there is a trap representation (negative zero) stored within the signed integer variable due to some test/product issue?
Not quite sure, but isn't it an UB to access it - so we may not expect any deterministic results from the test, incl. failure on different signed integer values?

Reply:
@v-klochkov was added comment, that explains why we cant compare all data type by `== operation`, also now `== operation` will be used for unsigned data type for all other will be used bitwise comparison
This change can be found in this commit: ddf461a2a5f4b538f7c331f2ca8b0a9970b18647

Reply:
"What if there is a trap representation (negative zero) stored within the signed integer variable"
Hmm, what do you mean by "negative zero for signed integer". Signed integer does not hold negative zero. Negative zero can be stored in FP value, but not in integer.
I may not understand some detail of implementation, please explain.

Reply:
Actually I was quite surprised as well. It's well-known that signed integral values overflow is UB. But the details are much more interesting.

SYCL 2020 works with C++17. Prior to C++20 there were 3 possible binary representations for signed integral types (see [basic.fundamental] in C++17):

- two's complement
- ones' complement
- signed magnitude

From the list above only the `two's complement` representation doesn't have a "negative zero" value, so every other representation has AFAIK a so-called "trap representation":

- ones' complement: all bits set to `1`
- signed magnitude: first bit set to `1`, the rest of bits set to `0`

One of the ways to retrieve such values is to have `INT_MAX +1` or `INT_MIN - 1`, but there is definitely much more ways to do so - for example, by using bitwise operators.

SYCL 2020 guarantees the binary representation on host side and device side are the same. But it doesn't have any additional guarantees on actual binary representation for the signed integral types.

C++17 is based on C11, which provides the definition of "trap representation" in p.6.2.6.1 (representation of types, general). Basically any variable may be initialized to store such representation without causing UB, but the value of the variable cannot be used until a proper value stored in it.

Reply:
Thank you for explanations.